File: experiments/016-LanguageComparisonRepeated/cells/java-gradle_1-Empty_free/2019-2c-parcial2-masked/20260917-061856-52134/analysis/acceptance-project/src/main/java/surveycrawler/SurveyCrawler.java
package surveycrawler;

import java.util.Random;

public class SurveyCrawler {

    private Point position;
    private CrawlerFacing facing;
    private final Sonar sonar;
    private final Random slideRandom;
    private CrawlerRunState runState;

    // instance creation

    public static SurveyCrawler atFacing(Point aPosition, String aFacingName) {
        return atFacingWithSonar(aPosition, aFacingName, aGroundPosition -> new FirmSand());
    }

    public static SurveyCrawler atFacingWithSonar(Point aPosition, String aFacingName, Sonar aSonar) {
        return atFacingWithSonarAndRandom(aPosition, aFacingName, aSonar, new Random());
    }

    public static SurveyCrawler atFacingWithSonarAndRandom(Point aPosition, String aFacingName, Sonar aSonar, Random aSlideRandom) {
        return new SurveyCrawler(aPosition, CrawlerFacing.facing(aFacingName), aSonar, aSlideRandom);
    }

    public static String invalidFacingErrorDescription() {
        return "Invalid facing";
    }

    public static int slideLimit() {
        return 10;
    }

    // exceptions

    public String invalidCommandErrorDescription() {
        return "Invalid command";
    }

    public String boulderFoundErrorDescription() {
        return "Boulder found";
    }

    public String siltFoundDuringGuardedRunErrorDescription() {
        return "Silt found during guarded run";
    }

    public String guardedRunAlreadyStartedErrorDescription() {
        return "Guarded run already started";
    }

    public String guardedRunNotStartedErrorDescription() {
        return "Guarded run not started";
    }

    public String guardedRunNotFinishedErrorDescription() {
        return "Guarded run not finished";
    }

    public void signalInvalidCommand() {
        throw new RuntimeException(invalidCommandErrorDescription());
    }

    public void signalBoulderFound() {
        throw new RuntimeException(boulderFoundErrorDescription());
    }

    public void signalSiltFoundDuringGuardedRun() {
        throw new RuntimeException(siltFoundDuringGuardedRunErrorDescription());
    }

    public void signalGuardedRunAlreadyStarted() {
        throw new RuntimeException(guardedRunAlreadyStartedErrorDescription());
    }

    public void signalGuardedRunNotStarted() {
        throw new RuntimeException(guardedRunNotStartedErrorDescription());
    }

    public void signalGuardedRunNotFinished() {
        throw new RuntimeException(guardedRunNotFinishedErrorDescription());
    }

    // initialization

    private SurveyCrawler(Point aPosition, CrawlerFacing aFacing, Sonar aSonar, Random aSlideRandom) {
        position = aPosition;
        facing = aFacing;
        sonar = aSonar;
        slideRandom = aSlideRandom;
        runState = new NormalRun();
    }

    // facing

    public void faceRight() {
        facing = new CrawlerFacingRight();
    }

    public void faceUp() {
        facing = new CrawlerFacingUp();
    }

    public void faceDown() {
        facing = new CrawlerFacingDown();
    }

    public void faceLeft() {
        facing = new CrawlerFacingLeft();
    }

    public void turnCounterClockwise() {
        groundUnderCrawler().turnCrawler(this, () -> facing.turnCounterClockwise(this));
    }

    public void turnClockwise() {
        groundUnderCrawler().turnCrawler(this, () -> facing.turnClockwise(this));
    }

    public void turnDoing(Runnable aTurn) {
        CrawlerFacing previousFacing = facing;
        aTurn.run();
        runState.registerUndo(() -> facing = previousFacing);
    }

    // testing

    public boolean isAtFacing(Point aPosition, String aFacingName) {
        return position.equals(aPosition) && facing.isFacing(aFacingName);
    }

    public boolean isRetreatCommand(char aCommand) {
        return aCommand == 't';
    }

    public boolean isAdvanceCommand(char aCommand) {
        return aCommand == 'a';
    }

    public boolean isTurnCounterClockwiseCommand(char aCommand) {
        return aCommand == 'g';
    }

    public boolean isTurnClockwiseCommand(char aCommand) {
        return aCommand == 'h';
    }

    public boolean isStartGuardedRunCommand(char aCommand) {
        return aCommand == '(';
    }

    public boolean isFinishGuardedRunCommand(char aCommand) {
        return aCommand == ')';
    }

    public boolean isRepeatableCommand(char aCommand) {
        return isAdvanceCommand(aCommand)
            || isRetreatCommand(aCommand)
            || isTurnClockwiseCommand(aCommand)
            || isTurnCounterClockwiseCommand(aCommand);
    }

    // moving

    public void retreat() {
        facing.retreat(this);
    }

    public void moveRight() {
        moveTowards(new Point(1, 0));
    }

    public void advance() {
        facing.advance(this);
    }

    public void moveUp() {
        moveTowards(new Point(0, 1));
    }

    public void moveDown() {
        moveTowards(new Point(0, -1));
    }

    public void moveLeft() {
        moveTowards(new Point(-1, 0));
    }

    public void moveTowards(Point aDirection) {
        sonar.groundAt(position.plus(aDirection)).moveCrawlerTowards(this, aDirection);
    }

    public void slideTowards(Point aDirection) {
        displaceBy(aDirection.times(slideRandom.nextInt(slideLimit()) + 1));
        runState.siltReached(this);
    }

    public void displaceBy(Point aDisplacement) {
        position = position.plus(aDisplacement);
        runState.registerUndo(() -> position = position.plus(aDisplacement.negated()));
    }

    private Ground groundUnderCrawler() {
        return sonar.groundAt(position);
    }

    // guarded run

    public void startGuardedRun() {
        runState.startGuardedRun(this);
    }

    public void finishGuardedRun() {
        runState.finishGuardedRun(this);
    }

    public void changeRunStateTo(CrawlerRunState aRunState) {
        runState = aRunState;
    }

    // command processing

    public void process(String aSequenceOfCommands) {
        try {
            processAll(aSequenceOfCommands);
        } catch (RuntimeException anError) {
            runState.abort(this);
            throw anError;
        }
    }

    private void processAll(String aSequenceOfCommands) {
        char previousCommand = ' ';
        for (char aCommand : aSequenceOfCommands.toCharArray()) {
            if (Character.isDigit(aCommand)) {
                repeatCommand(previousCommand, Character.digit(aCommand, 10) + 2);
            } else {
                processCommand(aCommand);
            }
            previousCommand = aCommand;
        }
        runState.finishSequence(this);
    }

    private void repeatCommand(char aCommand, int aNumberOfTimes) {
        if (!isRepeatableCommand(aCommand)) {
            signalInvalidCommand();
        }
        for (int i = 0; i < aNumberOfTimes; i++) {
            processCommand(aCommand);
        }
    }

    public void processCommand(char aCommand) {
        if (isAdvanceCommand(aCommand)) {
            advance();
            return;
        }
        if (isRetreatCommand(aCommand)) {
            retreat();
            return;
        }
        if (isTurnClockwiseCommand(aCommand)) {
            turnClockwise();
            return;
        }
        if (isTurnCounterClockwiseCommand(aCommand)) {
            turnCounterClockwise();
            return;
        }
        if (isStartGuardedRunCommand(aCommand)) {
            startGuardedRun();
            return;
        }
        if (isFinishGuardedRunCommand(aCommand)) {
            finishGuardedRun();
            return;
        }

        signalInvalidCommand();
    }
}

File: experiments/020-PythonVsJava/cells/python-pytest_1-Empty_free/2019-2c-parcial2-masked/20260921-210220-97414/analysis/own-tests-project/src/surveycrawler/survey_crawler.py
from surveycrawler.crawler_facing import CrawlerFacing
from surveycrawler.crawler_facing_right import CrawlerFacingRight
from surveycrawler.crawler_facing_up import CrawlerFacingUp
from surveycrawler.crawler_facing_down import CrawlerFacingDown
from surveycrawler.crawler_facing_left import CrawlerFacingLeft
from surveycrawler.firm_sand_sonar import FirmSandSonar
from surveycrawler.normal_run import NormalRun


class SurveyCrawler:

    # instance creation

    @classmethod
    def at_facing(cls, a_position, a_facing_name):
        return cls.at_facing_with_sonar(a_position, a_facing_name, FirmSandSonar())

    @classmethod
    def at_facing_with_sonar(cls, a_position, a_facing_name, a_sonar):
        return cls(a_position, CrawlerFacing.facing(a_facing_name), a_sonar)

    @classmethod
    def invalid_facing_error_description(cls):
        return "Invalid facing"

    # exceptions

    @classmethod
    def invalid_command_error_description(cls):
        return "Invalid command"

    def signal_invalid_command(self):
        raise RuntimeError(self.invalid_command_error_description())

    # initialization

    def __init__(self, a_position, a_facing, a_sonar):
        self._position = a_position
        self._facing = a_facing
        self._sonar = a_sonar
        self._run = NormalRun(self)

    # facing

    def face_right(self):
        self._facing = CrawlerFacingRight()

    def face_up(self):
        self._facing = CrawlerFacingUp()

    def face_down(self):
        self._facing = CrawlerFacingDown()

    def face_left(self):
        self._facing = CrawlerFacingLeft()

    def turn_counter_clockwise(self):
        self.assert_can_turn()
        self.turn_counter_clockwise_ignoring_ground()

    def turn_clockwise(self):
        self.assert_can_turn()
        self.turn_clockwise_ignoring_ground()

    def turn_counter_clockwise_ignoring_ground(self):
        self._facing.turn_counter_clockwise(self)

    def turn_clockwise_ignoring_ground(self):
        self._facing.turn_clockwise(self)

    # testing

    def is_at_facing(self, a_position, a_facing_name):
        return self._position == a_position and self._facing.is_facing(a_facing_name)

    def assert_can_turn(self):
        self.current_ground().assert_can_be_occupied()

    # ground detection

    def ground_at(self, a_position):
        return self._sonar.ground_at(a_position)

    def current_ground(self):
        return self.ground_at(self._position)

    # moving

    def retreat(self):
        self._facing.retreat(self)

    def advance(self):
        self._facing.advance(self)

    def move_towards(self, a_direction):
        self.ground_at(self._position.plus(a_direction)).move_crawler_towards(self, a_direction)

    def move_cells_towards(self, a_number_of_cells, a_direction):
        self._position = self._position.plus(a_direction.times(a_number_of_cells))

    def move_one_cell_forward(self):
        self.move_cells_towards(1, self._facing.direction())

    def move_one_cell_backward(self):
        self.move_cells_towards(1, self._facing.opposite_direction())

    def slide_towards(self, a_direction, a_silt):
        self._run.slide_towards(a_direction, a_silt)

    # command processing

    def process(self, a_sequence_of_commands):
        self.change_run_to(NormalRun(self))
        try:
            for a_character in a_sequence_of_commands:
                self._run.process(a_character)
            self._run.finish()
        finally:
            self.change_run_to(NormalRun(self))

    def change_run_to(self, a_run):
        self._run = a_run
